### DISCUSSION TITLE
ModuleNotFoundError running 'flask db migrate' after creating a shared models directory

### DISCUSSION THREAD

**Discussion by blacksteel1288** [miguelgrinberg/Flask-Migrate#554] [Q&A]
When I run a `flask db migrate`, I'm getting a `ModuleNotFoundError: No module named 'models' error`, that the `app/__init__.py` references.

`flask db migrate` worked previously with this app, but I recently consolidated my models into a single shared folder instead of having individual `models.py` files for each app (`api` and `web`)  that were getting out of sync.

I believe the issue has to do with setting the `PYTHONPATH` variable in some way before running the command, because when I run the app in PyCharm there are no issues.  I've specified the shared models directory as a content root in PyCharm which also adds it automatically to `PYTHONPATH`.  I also have no issues running the app in a docker container where I set `ENV PYTHONPATH="/topleveldir"` in the Dockerfile.

How can I set the `PYTHONPATH` correctly for `'flask db migrate'` to work?

Here's my project structure:

```
/topleveldir
|
+-/common
|
+-/models
    |
    +__init.py__
    +model1.py
    +model2.py
    +...
|
+-/api
|
+-/web
     |
     +-/app
           |
           +__init.py__
     +-/migrations
     |
     +run.py
```

And the `ModuleNotFoundError `comes from how the `register_extensions(app)` function setup works, where it requires importing the User, Role objects. Specifically the `from models import User, Role` line.  Here's the general pattern (simplified) I followed for that:

```
from flask import Flask
from flask_security import Security, SQLAlchemyUserDatastore
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()
security = Security()

def register_extensions(app):
    db.init_app(app)
    from models import User, Role
    user_datastore = SQLAlchemyUserDatastore(db, User, Role)
    security.init_app(app, user_datastore)

def create_app():
    app = Flask(__name__)
    app.config.from_object('config')  # Load configuration
    register_extensions(app)
    register_blueprints(app)
    configure_database(app)
    return app
```

I have found that if I manually set `PYTHONPATH` to `"/full/path/to/topleveldir"` before I run the `'flask db migrate'` it does run without the module not found error, but I was hoping there is a way to automate this or config setting?

Thank you!

**Comment by miguelgrinberg** [comment]
I don't really understand how you expect this to work. It looks like you have moved your models outside of the Flask application, so you have to tell Python where these models are so that the import works.

In many cases you don't have to explicitly set `PYTHONPATH` because the models can be imported from the current directory, which in most cases is automatically included in the path. In this case you are probably starting your Flask app from a directory other than the top level one, so doing `from models import ...` cannot be resolved.

If you don't want to explicitly set the import path, then run your app from the top level directory and then the import should work.

**Reply by blacksteel1288** [reply]
Thanks Miguel. Your answer was helpful in that it sent me in the right direction for a solution.

For background, there are two separate apps that are sharing this repo, api and web. Over time, I hope to combine them further, but for now they are separate, only sharing models and a few common utilities.

For the solution, I found a way to handle this issue inside of PyCharm. Since I'm able to run the apps (web and api) locally on PyCharm's run configurations, where it takes care of setting PYTHONPATH based on selecting the sources roots, it makes sense that I should be able to create a run configuration for doing a flask db migrate also. In those configurations, PYTHONPATH and other environment variables can be set and saved.

I just tested this and it seems to work as expected. All set.
